File: packages/core-java/src/main/java/com/openpasskey/core/Base64Url.java
package com.openpasskey.core;

import java.util.Base64;

/**
 * Base64url encoding/decoding utilities (no padding).
 */
public final class Base64Url {
    private Base64Url() {}

    public static byte[] decode(String input) {
        return Base64.getUrlDecoder().decode(input);
    }

    public static String encode(byte[] data) {
        return Base64.getUrlEncoder().withoutPadding().encodeToString(data);
    }
}

File: packages/server-py/open_passkey_server/base64url.py
"""Base64url encoding/decoding helpers."""

from base64 import urlsafe_b64decode, urlsafe_b64encode


def b64url_encode(data: bytes) -> str:
    return urlsafe_b64encode(data).rstrip(b"=").decode()


def b64url_decode(s: str) -> bytes:
    s += "=" * (-len(s) % 4)
    return urlsafe_b64decode(s)
